src/lib/network_pool/pool_sink.ml

```diff
@@ -0,0 +1,201 @@
+        ; trace_label : string
+        }
+        -> t
+    | Void
+
+  let on_overflow ~unwrap logger m =
+    match unwrap m with
+    | env, cb ->
+        Mina_metrics.(
+          Counter.inc_one Pipe.Drop_on_overflow.verified_network_pool_diffs) ;
+        let diff = Envelope.Incoming.data env in
+        [%log' warn logger] "Dropping verified diff $diff due to pipe overflow"
+          ~metadata:[ ("diff", Diff.verified_to_yojson diff) ] ;
+        BC.drop Diff.empty (Diff.reject_overloaded_diff diff) cb
+
+  (* keep logger in type *)
+  let verify_impl ~logger ~trace_label resource_pool rl env cb :
+      Diff.verified Envelope.Incoming.t option Deferred.t =
+    O1trace.trace_recurring (trace_label ^ "_verification") (fun () ->
+        if BC.is_expired cb then Deferred.return None
+        else
+          let summary = `String (Diff.summary @@ Envelope.Incoming.data env) in
+          [%log' debug logger] "Verifying $diff from $sender"
+            ~metadata:
+              [ ("diff", summary)
+              ; ("sender", Envelope.Sender.to_yojson env.sender)
+              ] ;
+          match
+            Rate_limiter.add rl env.sender ~now:(Time.now ())
+              ~score:(Diff.score env.data)
+          with
+          | `Capacity_exceeded ->
+              [%log' debug logger]
+                ~metadata:
+                  [ ("sender", Envelope.Sender.to_yojson env.sender)
+                  ; ("diff", summary)
+                  ]
+                "exceeded capacity from $sender" ;
+              BC.error (Error.of_string "exceeded capacity") cb
+              >>| fun _ -> None
+          | `Within_capacity -> (
+              match%bind Diff.verify resource_pool env with
+              | Error err ->
+                  [%log' debug logger]
+                    "Refusing to rebroadcast $diff. Verification error: $error"
+                    ~metadata:
+                      [ ("diff", summary)
+                      ; ("error", Error_json.error_to_yojson err)
+                      ] ;
+                  (*reject incoming messages*)
+                  BC.error err cb >>| fun _ -> None
+              | Ok verified_diff ->
+                  [%log' debug logger] "Verified diff: $verified_diff"
+                    ~metadata:
+                      [ ( "verified_diff"
+                        , Diff.verified_to_yojson
+                          @@ Envelope.Incoming.data verified_diff )
+                      ; ( "sender"
+                        , Envelope.Sender.to_yojson
+                          @@ Envelope.Incoming.sender verified_diff )
+                      ] ;
+                  Deferred.return (Some verified_diff) ))
+
+  let push t (msg, cb) =
+    match t with
+    | Sink { writer = w; logger; rate_limiter = rl; pool; wrap; trace_label }
+      -> (
+        let env' = Msg.convert msg in
+        let cb' = Msg.convert_callback cb in
+        match%bind verify_impl ~logger ~trace_label pool rl env' cb' with
```

Comment:
> blocking on an RPC call to another process

Could you describe why is it necessary?

Reply:
Because we will not process any other pubsub messages while this message is being verified.

Reply:
@nholland94 what do you think of having strict pipe contain unverified network pool diffs, not verified?

I would still run rate limiting check before writing to strict pipe.

This seems like the clearest solution to problem.
Or I could make it two strict pipes: one for unverified and then another for verified, but I can not find a justification for such a complicated model.

Do we want to limit flow of verified network pool diffs specifically, not just any network pool diffs?

Reply:
I think we would want 2 strict pipes. I think we can just do one pipe and safely spawn `don't_wait_for` background jobs for verifications. It's safe to have a bunch of background jobs for verification because most of the work involved in verifying these diffs is spent in waiting for I/O (communicating with the snark verifier). We can use a `Throttle` to limit the number of active verification jobs, and each job will write into the pipe if verification is successful.